pkg/releaseindex/release_index.go

```diff
@@ -0,0 +1,163 @@
+package releaseindex
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/versionbundle"
+	"github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+	yaml "gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	// GithubToken is a Github token which is authorized to read from the
+	// installations repository.
+	GithubToken string
+	Logger      micrologger.Logger
+}
+
+type ReleaseIndex struct {
+	githubToken string
+	logger      micrologger.Logger
+
+	cache    map[Provider][]versionbundle.IndexRelease
+	cacheMux *sync.RWMutex
+}
+
+func New(config Config) (*ReleaseIndex, error) {
+	if config.GithubToken == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GithubToken must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &ReleaseIndex{
+		githubToken: config.GithubToken,
+		logger:      config.Logger,
+
+		cache:    map[Provider][]versionbundle.IndexRelease{},
+		cacheMux: &sync.RWMutex{},
+	}
+
+	return r, nil
+}
+
+// GetVersion returns a component version for a given type and provider.
+//
+// NOTE: Release index is cached for a provider after first call. This makes it
+// cheap to call this method multiple times with the same provider. But changes
+// to the installations repository will not be visible.
+func (r *ReleaseIndex) GetVersion(ctx context.Context, provider Provider, versionType VersionType, component Component) (string, error) {
+	index, err := r.getIndex(ctx, provider)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	version, err := getVersion(index, versionType, component)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return version, nil
+}
+
+func (r *ReleaseIndex) getFileContent(ctx context.Context, owner, repo, path string) (string, error) {
+	var client *github.Client
+	{
+		t := &oauth2.Token{
+			AccessToken: r.githubToken,
+		}
+
+		sts := oauth2.StaticTokenSource(t)
+		tc := oauth2.NewClient(ctx, sts)
+
+		client = github.NewClient(tc)
+	}
+
+	opt := &github.RepositoryContentGetOptions{}
+	repoContent, _, _, err := client.Repositories.GetContents(ctx, owner, repo, path, opt)
+
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	content, err := repoContent.GetContent()
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return content, nil
+}
+
+func (r *ReleaseIndex) getIndex(ctx context.Context, provider Provider) ([]versionbundle.IndexRelease, error) {
```

Comment:
Not sure how I feel about this, I kind of feel like this should be part of a general lib.
We already get the `releaseIndex` from `installations` in `opsctl` and we could also need it in other upcoming stories like the app-catalog thing. I would like it if we had a shared codebase for that.

See osctl:https://github.com/giantswarm/opsctl/blob/master/service/installation/repo/repo.go#L270

Reply:
Do you have an idea where should that land? `versionbundle` repo?

Reply:
Versionbundle repo sounds fine to me.

Reply:
I checked that repo and I think there are more things to get straight. In general version bundle will be obsolete when we get into the versioning of operators story. I will add a note there to not forget about that.

Reply:
I added TODO here https://github.com/giantswarm/e2esetup/pull/12/commits/aa5a48ee80be279b57f97b8bb75a1f05ba47ed93.

We'll revisit that during versioning of operators.